### PR TITLE
Krita 5.0.2 and runtime 21.08

### LINKED ISSUE #38: Support fcitx im module
Im module need to be bundled with in the application as a plugin to support input methods.

Actually I'd suggest krita to switch to KDE's runtime instead of freedesktop one since it already has fcitx bundled (And also other dependencies that might be useful).

If you want to add it to the build file KDE's manifest can be used as a reference.
https://invent.kde.org/packaging/flatpak-kde-runtime/-/blob/qt5.15lts/org.kde.Sdk.json.in#L1012

### LINKED ISSUE #48: Krita 5.0
https://krita.org/en/item/krita-5-0-released/

Hopefully, this can be based on KDE Platform?

### PR DESCRIPTION
**Pull request by vulongm**
* Bumped version to Krita 5.0.2 (resolves #48)
* Added and updated all dependencies required by Krita 5.0.0 from the [repo](https://invent.kde.org/graphics/krita/-/tree/master/3rdparty) (resolves #38, #47)
* Qt 5.12.12 with additional patches to fix build on runtime 21.08 (resolves #45)

### CODE REVIEW

File: org.kde.krita.yaml [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+modules:
+  - name: png
+    buildsystem: cmake-ninja
+    builddir: true
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DPNG_TESTS=OFF
+    sources:
+      - type: archive
+        url: https://download.sourceforge.net/libpng/libpng-1.6.37.tar.gz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:42]: The runtime has this exact version

File: org.kde.krita.yaml [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+        sha256: daeb2620d829575513e35fecc83f0d3791a620b9b93d800b763542ece9390fb4
+  - name: tiff
+    buildsystem: cmake-ninja
+    builddir: true
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+    sources:
+      - type: archive
+        url: https://download.osgeo.org/libtiff/tiff-4.2.0.tar.gz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:51]: the runtime has version 4.3.0 of libtiff.

File: org.kde.krita.yaml [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+          - patches/tiff/0001-Enable-writing-Photoshop-blobs.patch
+  - name: jpeg
+    buildsystem: cmake-ninja
+    builddir: true
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DWITH_SIMD=OFF
+    sources:
+      - type: archive
+        url: https://downloads.sourceforge.net/project/libjpeg-turbo/2.0.6/libjpeg-turbo-2.0.6.tar.gz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:65]: the runtime has libjpeg-turbo 2.1.2.

File: org.kde.krita.yaml [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DEXIV2_BUILD_SAMPLES=OFF
+      - -DEXIV2_ENABLE_NLS=OFF
+      - -DEXIV2_BUILD_EXIV2_COMMAND=OFF
+    cleanup:
+      - /bin
+    sources:
+      - type: archive
+        url: https://mirrors.ocf.berkeley.edu/kde-applicationdata/krita/build/dependencies/exiv2-0.27.3-Source.tar.gz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:120]: 0.27.5 has been out for a bit. and it allow dropping the patch below AFAIK (breakage on aarch64)

File: org.kde.krita.yaml [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+      - -DBUILD_GTK_TESTS=OFF
+      - -DBUILD_QT5_TESTS=FALSE
+      - -DBUILD_CPP_TESTS=FALSE
+      - -DENABLE_UTILS=FALSE
+      - -DENABLE_GLIB=FALSE
+      - -DENABLE_LIBOPENJPEG=none
+      - -DWITH_NSS3=OFF
+    sources:
+      - type: archive
+        url: https://poppler.freedesktop.org/poppler-21.02.0.tar.xz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:540]: you should update to 21.12.0. It fixes a few CVE

File: org.kde.krita.yaml [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+        sha256: b45183d9a51a65dacaed33726587a1604d09c77adffea9f02735085e9335053d
+  - name: webp
+    buildsystem: cmake-ninja
+    builddir: true
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DBUILD_SHARED_LIBS=OFF
+    sources:
+      - type: archive
+        url: https://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.2.0.tar.gz
```
Comment (hfiguiere) [inline org.kde.krita.yaml:621]: libwebp 1.2.1 is in the runtime.

**Comment by hrstwn** [comment]
I have tested this and it works.

**Comment by joshua-stone** [comment]
@vulongm @hrstwn I've done some testing on the latest Krita build as well, and can confirm that it works. Krita launches and runs without obvious errors, having tried the following:

* saving new documents
* changing brush settings
* re-opening recent documents
* saving files
* exporting files
* drawing with a tablet [(model is One by Wacom)](https://www.wacom.com/en-us/products/pen-tablets/one-by-wacom)

CLI output also doesn't appear to be normal:

```
$ flatpak install --user https://dl.flathub.org/build-repo/70230/org.kde.krita.flatpakref
The application org.kde.krita depends on runtimes from:
  https://dl.flathub.org/repo/
Configure this as new remote 'flathub' [Y/n]: n

org.kde.krita permissions:
    ipc     network     pulseaudio     x11     dri     file access [1]     system dbus access [2]

    [1] host, xdg-run/gvfs
    [2] org.freedesktop.ColorManager


        ID                             Branch          Op          Remote                Download
 1. [✓] org.kde.krita.Locale           test            i           krita-origin            8.1 kB / 11.5 MB
 2. [✓] org.kde.krita                  test            i           krita-origin          156.6 MB / 156.5 MB

$ flatpak run org.kde.krita 
Qt: Session management error: None of the authentication protocols specified are supported
Qt: Session management error: None of the authentication protocols specified are supported
Qt: Session management error: None of the authentication protocols specified are supported
Gtk-Message: 09:27:06.278: Failed to load module "canberra-gtk-module"
Gtk-Message: 09:27:06.278: Failed to load module "canberra-gtk-module"
Qt: Session management error: None of the authentication protocols specified are supported
QObject::startTimer: Timers cannot have negative intervals
/app/lib/krita-python-libs/krita added to PYTHONPATH
```

**Comment by vulongm** [comment]
> CLI output also doesn't appear to be normal

Are you able to replicate that on the official Krita 5 AppImage as well? I did notice ``QObject::startTimer: Timers cannot have negative intervals`` appears on both the Flatpak and AppImage build for me testing it out, but not sure what is causing the other errors.

**Comment by hfiguiere** [comment]
the first 6 message I have seen them with almost every flatpak Qt app. (the canberra-gtk-module with even more)

**Comment by joshua-stone** [comment]
@vulongm This is what I get with the official appimage:

```
$  ./krita-5.0.0-x86_64.appimage 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
QObject::startTimer: Timers cannot have negative intervals
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
/tmp/.mount_krita-Aq3tJe/usr/lib/krita-python-libs/krita added to PYTHONPATH
```

**Comment by joshua-stone** [comment]
@vulongm Oh sorry, I didn't realize what I typed. I meant that the CLI output appears to be normal. Sorry for the miscommunication!

**Comment by VarLad** [comment]
@vulongm Can't test because
> flatpak install https://dl.flathub.org/build-repo/70230/org.kde.krita.flatpakref
error: Can't load uri https://dl.flathub.org/build-repo/70230/org.kde.krita.flatpakref: Server returned status 404: Not Found

**Comment by hfiguiere** [comment]
Test build expired

bot, build

**Comment by jsj1027** [comment]
Are there any blocking issues left?

**Comment by hrstwn** [comment]
Sorry but any chance to have this merged?

**Comment by vulongm** [comment]
bot, build

**Comment by vulongm** [comment]
Any ideas what is causing the build here to fail? I'm not sure if the appdata file change would have an impact on that particular part of the build process either where it is failing.

> stripping /srv/buildbot/worker/build-x86_64-5/build/.flatpak-builder/rofiles/rofiles-KovXOm/files/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug to /srv/buildbot/worker/build-x86_64-5/build/.flatpak-builder/rofiles/rofiles-KovXOm/files/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug
FB: Running: eu-strip --remove-comment --reloc-debug-sections -f /srv/buildbot/worker/build-x86_64-5/build/.flatpak-builder/rofiles/rofiles-KovXOm/files/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug -F /app/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug /srv/buildbot/worker/build-x86_64-5/build/.flatpak-builder/rofiles/rofiles-KovXOm/files/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug
eu-strip: cannot open '/srv/buildbot/worker/build-x86_64-5/build/.flatpak-builder/rofiles/rofiles-KovXOm/files/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/debug/lib/libQt5EglFSDeviceIntegration.so.5.12.12.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug.debug': File name too long
]2;flatpak-builder: Cleanup openexrError: module openexr: Child process exited with code 1

**Comment by hfiguiere** [comment]
This is related to https://github.com/flatpak/flatpak-builder/issues/429

**Comment by vulongm** [comment]
bot, build

**Comment by vulongm** [comment]
bot, cancel

**Comment by hfiguiere** [comment]
We hope to have a fix deployed soon for this flatpak-builder bug causing build failures

**Comment by hfiguiere** [comment]
bot, build

**Comment by hfiguiere** [comment]
retriggered build as the fix has been deployed.

**Comment by hfiguiere** [comment]
not sure why the x86_64 build failed on the screenshot. Maybe a glitch with the cdn serving them?

**Comment by joshua-stone** [comment]
Is that screenshot not available anymore?

https://cdn.kde.org/screenshots/krita/2018-03-17_screenshot_002.png

It doesn't appear to be on the file server, or maybe it was renamed/moved a few days ago:

https://cdn.kde.org/screenshots/krita/

**Comment by VarLad** [comment]
There are other screenshots on the page that can be used https://cdn.kde.org/screenshots/krita/

**Comment by vulongm** [comment]
Krita changed the screenshot URLs recently but looks like that one wasn't changed yet. Check the bug report here https://bugs.kde.org/show_bug.cgi?id=447520.

**Comment by hrstwn** [comment]
the build seems to runs fine.

Here's what's shown on terminal:
`$ flatpak run org.kde.krita
Gtk-Message: 17:08:59.073: Failed to load module "canberra-gtk-module"
Gtk-Message: 17:08:59.074: Failed to load module "canberra-gtk-module"
QIBusPlatformInputContext: invalid portal bus.
Qt: Session management error: None of the authentication protocols specified are supported
QIBusPlatformInputContext: invalid portal bus.
Qt: Session management error: None of the authentication protocols specified are supported
QIBusPlatformInputContext: invalid portal bus.
Qt: Session management error: None of the authentication protocols specified are supported
QIBusPlatformInputContext: invalid portal bus.
Qt: Session management error: None of the authentication protocols specified are supported
QApplication: invalid style override passed, ignoring it.
    Available styles: Windows, Fusion
krita.general: ERROR: unable to decode abr format version  23364 (subver  0 )
QObject::startTimer: Timers cannot have negative intervals
/app/lib/krita-python-libs/krita added to PYTHONPATH
`

This line seems to not appear on previous builds:
`krita.general: ERROR: unable to decode abr format version  23364 (subver  0 )`

**Comment by vulongm** [comment]
> This line seems to not appear on previous builds: `krita.general: ERROR: unable to decode abr format version 23364 (subver 0 )`

That's strange, not able to replicate that here on the current flatpak build or AppImage for 5.0.2.

**Comment by hrstwn** [comment]
I tried to install the build on my other computer and the message did not appear there. 
Maybe it's just my laptop. But both seem to run perfectly. 
I tried importing a .abr brush to check and it works fine.

**Comment by vulongm** [comment]
@travier Any ideas when this can be reviewed/merged? The newest build seems to be functional, not noticing any issues so far.

**Comment by travier** [comment]
Taking a look. Sorry I had not seen this one earlier. Feel free to ping me for reviews.

**Comment by travier** [comment]
Thanks for the work here.
This is a large PR and I don't have the full history here so here are a few questions:
- Shouldn't we use the official mirrors for frameworks? https://download.kde.org/Attic/frameworks/5.76 should have them.
- Are https://mirrors.ocf.berkeley.edu/kde-applicationdata/krita/build/dependencies/flatpak/* available somewhere on the official mirrors? Should we use Git clones + fixed tag instead?
- Same for https://mirrors.ocf.berkeley.edu/kde/stable/krita/5.0.2/krita-5.0.2.tar.gz, should probably be from the official mirror?
- That's a lot of patches. Are they all from Krita repo?

**Comment by vulongm** [comment]
> Shouldn't we use the official mirrors for frameworks? https://download.kde.org/Attic/frameworks/5.76 should have them.

When download from KDE mirrors at download.kde.org flatpak would run into an HTTP2 framing error resulting in the build failing. Using the mirror seemed to not have those issues. I'm not sure if there is another workaround where we can avoid those errors on the official mirrors.

> Are https://mirrors.ocf.berkeley.edu/kde-applicationdata/krita/build/dependencies/flatpak/* available somewhere on the official mirrors? Should we use Git clones + fixed tag instead?

They are available at https://files.kde.org/krita/build/dependencies/flatpak/ on the official mirror. They come packaged with dependencies needed for the offline install since cargo can't grab additional dependencies in flatpak building process. 

> Same for https://mirrors.ocf.berkeley.edu/kde/stable/krita/5.0.2/krita-5.0.2.tar.gz, should probably be from the official mirror?

It might be possible to try moving the dependencies mentioned in Q2/Q3 to the official mirrors. I mostly notice the errors occurring for the KDE Frameworks download from Q1 so not sure if that issue is present potentially with the Krita/flatpak dependencies.

> That's a lot of patches. Are they all from Krita repo?

Most of them are from the Krita repo at https://invent.kde.org/graphics/krita/-/tree/master/3rdparty.

Outside that "qt/declarative/Additional-GCC11-Fixes.patch" was added since the other patches seemed to be missing more GCC11 related changes needed to build on runtime 21.08.

The pystring CMake file "patches/pystring/CMakeLists.txt" was from OpenColorIO "share/cmake/projects/Buildpystring.cmake" which is one of Krita's dependencies, and pystring is a dependency of OpenColorIO.

There's also the AppData patch at "patches/krita/AppData.patch" which backports some changes made upstream that wasn't in 5.0.2 for that file. Along with fixing a typo "brusheditor.png" to "brush_editor.png" and removed an extra "xx" character that was causing the build to fail for the Krita Artists community link x-test lang caption.

**Comment by travier** [comment]
OK, let's give this one a chance. A lot of things are not ideal here but not wrong either so we will try to fix them as we go. Thanks!

**Pull request merged**